ldap/servers/plugins/replication/cl5_api.c

```diff
@@ -1705,6 +1739,226 @@ cl5DBData2Entry(const char *data, PRUint32 len __attribute__((unused)), CL5Entry
+            } else {
+                clrc = CL5_DB_ERROR;
+            }
+            break;
+    }
+    if (errmsg) {
+        slapi_log_err(loglvl, repl_plugin_name_cl,
+                      "%s on changelog %s; db error - %d %s\n",
+                      errmsg, cldb->ident, rc, dblayer_strerror(rc));
+    }
+    return clrc;
+}
+
+/* Iterate over a changelog cursor */
+int
+_cl5Iterate(cldb_Handle *cldb, dbi_iterate_cb_t *action_cb, DBLCI_CTX *dblcictx, PRBool readonly)
+{
+    dbi_txn_t *txnid = NULL;
+    int rc = CL5_DB_RETRY;
+    int nbtries = 0;
+    CSN *startcsn = NULL;
+
+    dblcictx->finished = PR_FALSE;
+    dblcictx->cldb = cldb;
+    while ( !slapi_is_shutting_down() &&
+            ((rc == CL5_SUCCESS && dblcictx->finished == PR_FALSE) ||
+             (rc == CL5_DB_RETRY && nbtries < MAX_RETRIES))) {
+        nbtries++;
+        dblcictx->changed.nb = 0;
+        dblcictx->seen.nb = 0;
+        if (rc == CL5_SUCCESS) {
+            /* action_cb have decided to abort/commit the txn
+             * then go on with the iteration
+             */
+            startcsn = &dblcictx->csn;
+        }
+        if (startcsn) {
+            /* Save startcsn to be able to retry at the same place */
+            dblcictx->startcsn = *startcsn;
+            startcsn = &dblcictx->startcsn;
+        }
+        if (!readonly || dblayer_is_lmdb(cldb->be)) {
+            /*
+             * if write operation are performed or if lmdb is used, a transaction is needed.
+             * In bdb case DB txn lock accessed pages until the end of the transaction.
+             * in lmdb case the read or write lock is global and also held until the end
+             * of the transaction.
+             */
+            rc = TXN_BEGIN(cldb, NULL, &txnid, readonly);
+            if (rc != 0) {
+                rc = _cl5Dberror(cldb, rc, "_cl5Iterate - Failed to begin transaction");
+                continue;
+            }
+        } else {
+            /* read-only opertion on bdb are transactionless, so no reason to abort txn 
+             * after having seen some number of records 
+             */
+            dblcictx->seen.nbmax = 0;
+        }
+        /* create cursor */
+        rc = dblayer_new_cursor(cldb->be, cldb->db, txnid, &dblcictx->cursor);
+        if (rc != 0) {
+            rc = _cl5Dberror(cldb, rc, "_cl5Iterate - Failed to create cursor");
+            TXN_ABORT(cldb, txnid);
+            continue;
+        }
+        if (startcsn) {
+            char csnstr[CSN_STRSIZE] = "";
+            dbi_val_t startingkey = {0};
+            csn_as_string(startcsn, 0, csnstr);
+            dblayer_value_set(cldb->be, &startingkey, csnstr, CSN_STRSIZE);
+            rc = dblayer_cursor_iterate(&dblcictx->cursor, action_cb, &startingkey, dblcictx);
+        } else {
+            rc = dblayer_cursor_iterate(&dblcictx->cursor, action_cb, NULL, dblcictx);
+        }
+        if (dblcictx->seen.nbmax > 10 && rc == CL5_DB_LOCK_ERROR) {
```

Comment:
FYI: I agree with your first statement. That is why some time ago I open  [Issue 5679](https://github.com/389ds/389-ds-base/issues/5679) 
Anyway IMHO this code is now obsolete because the db lock monitoring thread should stop the server before bdb returns this error.

Reply:
But I disagree with your suggestion: We must not retry after decreasing the max number of seen records  
 for any database error